[CSL-1473] First draft for CBOR Sets

adinapoli-iohk opened the pull request:
See https://issues.serokell.io/issue/CSL-1473 for the full context.

We need such draft to be reviewed (possibly by @dcoutts) and then hosted publicly on GH (which is why I've put it inside the `specs` folder). Last but not least we would need to notify IANA and wait for the registration process to complete.

neongreen commented:
Shouldn't this spec go into a bit more detail regarding how sets have different semantics from arrays? E.g.

> ### Semantics
>
> Tag 258 can be applied to an array to indicate that the tagged
> object is a set. Sets should be handled similarly to maps: a set
> that has duplicate elements may be well-formed, but it is not
> valid. Other constraints on map keys ([Section 3.7]) apply to set
> elements as well.

[Section 3.7]: https://tools.ietf.org/html/rfc7049#section-3.7

adinapoli-iohk commented:
Happy to incorporate that @neongreen !

dcoutts commented on specs/CBOR_SETS.md:
```diff
@@ -0,0 +1,29 @@
+
+Tag 258 can be applied to an array to indicate that the tagged object is a set. Sets should be handled
+similarly to maps: a set that has duplicate elements may be well-formed, but it is not valid.
+Other constraints on map keys [2] apply to set elements as well.
```
Is it worth being explicit here rather than only referring?

Using the same wording as the RFC for maps, as you've done, is sensible.

dcoutts commented:
Direct link for other people to see similar examples: https://www.iana.org/assignments/cbor-tags/cbor-tags.xhtml

adinapoli-iohk commented on specs/CBOR_SETS.md:
```diff
@@ -0,0 +1,29 @@
+    Tag: 258
+    Data item: array
+    Semantics: Set
+    Point of contact: Input Output HK <info@iohk.io>
```
Who is the person to enquire for this piece of information? (Sorry for this "spoonfeeding" request, but I'm still relatively new to the company structure)

adinapoli-iohk commented:
@dcoutts @neongreen I have pushed a second revision out. Apart from the typos which might lurk around, I have done a couple of changes:

* I have changed the email to `operations@iohk.io` because according to Costas is our best bet (the whole ops team will be notified, which I guess it's more than enough to drag our attention);

* I have "unrolled" Section 3.7 (constraints on map) and adapted it for sets.

* I have made a point (which I'm not 100% sure about, but I think it's the right direction) about the types of the items in a set.

* Included real authors ❤️

cabo commented:
FYI: I sent a few comments to the authors.  I would like to see a cleanup so that the spec text crisply says what a set is (a separate introduction is fine), and I'd like to see a clear definition of the canonical format of a set if that is important to you.  (I also could imagine going for specification required.)

adinapoli-iohk commented:
> One last thing is that perhaps we should think about what a really stable URL is going to be and post it there. It'd be annoying to have it move just because we do some re-organisation of the code base in the next 12 months. Perhaps a repo on it's own? Or plausibly we could host it with the cborg package implementation. Depends if IOHK see this as glory or a burden.

Good point! I'm open to recommendations. Looking at what [other people have done](https://github.com/lucas-clemente/cbor-specs/blob/master/id.md), I can see how the "self-contained-repo" option seems to be [fairly popular](https://britram.github.io/rains-prototype/)

adinapoli-iohk commented:
@dcoutts @cabo Ok, I think we are getting closer. This is what I have done:

* Expanded a little bit on what a set is from a mathematical point of view (I have cited Wolfram that should be decently authoritative, please shout if you think that's bit lame instead);

* Cleaned up a little bit the vague section on single vs multiple types as items of a set;

* Unrolled the sorting spec used for maps and adapted it for sets (we all agree that the current sorting scheme for maps is a bit surprising but nevertheless proposing a new one for sets would be even more confusing);

* Added an practical example.

Let me know what you guys think, hopefully this is a step forward!

adinapoli-iohk commented:
Thanks for the review, @dcoutts !

I'm now going to move this document in a separate repo, in order to give it a stable URL.

adinapoli-iohk commented:
Moved here: https://github.com/input-output-hk/cbor-sets-spec/blob/master/CBOR_SETS.md

cabo commented:
FYI: I have alerted the IETF CBOR WG to this spec, and have specifically asked whether we should encourage you to go for a two-byte tag.  I'm looking forward to some feedback now.

<https://mailarchive.ietf.org/arch/msg/cbor/CNm4-DjSKGqRdEh_Hn7lEtv4eu8>

adinapoli-iohk commented:
Ah, thank you very much, @cabo !

We save submitted the proposal to IANA, via [this form](https://www.iana.org/form/protocol-assignment) prior to reading this message 😞 
hopefully they will be able to interpolate your email with their final decision 😉 

(TL;DR we asked for tag 258, let's see what lies in store!)
